File: Week3/Python/ArrayProblems/ReversePairs.py
"""
    Given an integer array nums, return the number of reverse pairs in the array.
    A reverse pair is a pair (i, j) where:
    0 <= i < j < nums.length and nums[i] > 2 * nums[j]

    https://leetcode.com/problems/reverse-pairs/
"""


def reversePairs(nums):
    """ T(c) -> O(n^2), S(c) -> O(1) """
    count = 0

    for i in range(len(nums)):
        for j in range(i+1, len(nums)):
            if nums[i] > 2 * nums[j]:
                count += 1
                
    return count


def reversePairsEfficient(nums):
    """ T(c) -> O(nlogn), S(c) -> O(n) """

    # using merge sort
    return mergeSort(nums, 0, len(nums)-1)
    
def mergeSort(nums, low, high):
    count = 0
    
    if low < high:
        mid = (low+high) // 2
        count += mergeSort(nums, low, mid)
        count += mergeSort(nums, mid+1, high)
        count += countReversePairs(nums, low, mid, high)
        merge(nums, low, mid, high)
    
    return count

def countReversePairs(nums, low, mid, high):
    cnt = 0
    j = mid+1
    
    for i in range(low, mid+1):
        # check second part element with first part element
        while j <= high and nums[i] > 2* nums[j]:
            j += 1
    
        cnt += j - (mid+1)
    
    return cnt
        
def merge(nums, low, mid, high):
    l = low
    r = mid+1
    k = 0
    B = [0] * (high-low+1)        
    
    while l <= mid and r <= high:
        if nums[l] <= nums[r]:
            B[k] = nums[l]
            l += 1
        else:
            B[k] = nums[r]
            r += 1
        k += 1
    
    # elements present in left part
    while l <= mid:
        B[k] = nums[l]
        k += 1
        l += 1
    
    # elements present in right part
    while r <= high:
        B[k] = nums[r]
        k += 1
        r += 1
        
    # shift elements from B array to original array
    for i in range(len(B)):
        nums[i+low] = B[i]


arr = [1,3,2,3,1]
print(arr)
# The reverse pairs are:
# (1, 4) --> nums[1] = 3, nums[4] = 1, 3 > 2 * 1
# (3, 4) --> nums[3] = 3, nums[4] = 1, 3 > 2 * 1
print(reversePairs(arr))

arr = [2,4,3,5,1]
print(arr)
print(reversePairs(arr))

File: Week3/Java/ArrayProblem/ReversePairs.java
public class ReversePairs {

/*
 * Given an integer array nums, return the number of reverse pairs in the array.
 * A reverse pair is a pair (i, j) where: 0 <= i < j < nums.length and nums[i] > 2 * nums[j].
 * 
 * https://leetcode.com/problems/reverse-pairs/description/
 * T(c) -> O(n^2), S(c) -> O(1)
 */

    static int reversePairs(int[] nums) {
        int n = nums.length;
        int count = 0;
        for(int i=n-1;i>=0;i--)
        {
            for(int j=i+1;j<n;j++)
            {
                // convert to long, as integer can overflow after multiply
                if(nums[i] > (long)2*nums[j])
                {
                    count++;
                }
            }
        }
        return count;
    }


    /*
     * Using merge sort
     * T(c) -> O(nlogn) + O(n) + O(n) => O(2nlogn)
     * S(c) -> O(n) 
     */

    public static int reversePairsEfficient(int[] nums) {
        return mergeSort(nums, 0, nums.length-1);
    }
    
    static int mergeSort(int[] nums, int low, int high)
    {
        int cnt = 0;
        // run until one or no element left
        if(low<high)
        {
            int mid = (low+high)/2;
            // left part
            cnt += mergeSort(nums, low, mid);
            // right part
            cnt += mergeSort(nums, mid+1, high);
            // count reverse pairs
            cnt += countReversePairs(nums, low, mid, high);
            // merge both parts
            merge(nums, low, mid, high);
        }  
        return cnt;
    }
    
    static void merge(int[] nums, int low, int mid, int high)
    {
        int i = low;
        int j = mid+1;
        int k = 0;
        
        int temp[] = new int[high-low+1];  
        
        // both part elements present
        while(i<=mid && j<=high)
        {
            if(nums[i] <= nums[j])
            {
                temp[k++] = nums[i++];
            }
            else{
                temp[k++] = nums[j++];
            }
        }
        
        // remaining elements in first part
        while(i<=mid)
        {
            temp[k++] = nums[i++];
        }
        
        // remaining elements in second part
        while(j<=high)
        {
            temp[k++] = nums[j++];
        }
        
        // copy elements from temp array to original array
        for(int p=low;p<=high;p++)
        {
            nums[p] = temp[p-low];
        }
    }
    
    static int countReversePairs(int nums[], int low, int mid, int high)
    {
        int count = 0;
        int right = mid+1;
        for(int i=low; i<=mid; i++)
        {
            // check 2nd part elements with 1st part element
            while(right <= high && nums[i] > 2L*nums[right])
            {
                right++;
            }
            count+= right-(mid+1);
        }
        return count;
    } 


    public static void main(String[] args) {
        int arr1[] = {1,3,2,3,1};
        /* (1, 4) --> nums[1] = 3, nums[4] = 1, 3 > 2 * 1 
        (3, 4) --> nums[3] = 3, nums[4] = 1, 3 > 2 * 1 */
        System.out.println("Number of reverse pairs = "+ reversePairs(arr1));
        int arr2[] = {2147483647,2147483647,2147483647,2147483647,2147483647,2147483647};
        System.out.println("Number of reverse pairs = "+ reversePairsEfficient(arr2));
    }
    
}
